### PR TITLE
[SPARK-23831][SQL] Add org.apache.derby to IsolatedClientLoader

### PR DESCRIPTION
**Pull request by wangyum**
## What changes were proposed in this pull request?

Add `org.apache.derby` to `IsolatedClientLoader`, otherwise it may throw an exception:
```scala
...
[info] Cause: java.sql.SQLException: Failed to start database 'metastore_db' with class loader org.apache.spark.sql.hive.client.IsolatedClientLoader$$anon$1@2439ab23, see the next exception for details.
[info] at org.apache.derby.impl.jdbc.SQLExceptionFactory.getSQLException(Unknown Source)
[info] at org.apache.derby.impl.jdbc.SQLExceptionFactory.getSQLException(Unknown Source)
[info] at org.apache.derby.impl.jdbc.Util.seeNextException(Unknown Source)
[info] at org.apache.derby.impl.jdbc.EmbedConnection.bootDatabase(Unknown Source)
[info] at org.apache.derby.impl.jdbc.EmbedConnection.<init>(Unknown Source)
[info] at org.apache.derby.jdbc.InternalDriver$1.run(Unknown Source)
...
```

## How was this patch tested?

unit tests and manual tests

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 1]
```diff
@@ -188,6 +188,9 @@ private[hive] class IsolatedClientLoader(
     (name.startsWith("com.google") && !name.startsWith("com.google.cloud")) ||
     name.startsWith("java.lang.") ||
     name.startsWith("java.net") ||
+    name.startsWith("com.sun.") ||
+    name.startsWith("sun.reflect.") ||
```
Comment (felixcheung) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: these two lines above are really broad, is this intentional?

**Comment by wangyum** [comment]
cc @jerryshao

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala (hunk 1 above)
Comment (gatorsmile) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: Do not add them unless we have to do it.

**Comment by gatorsmile** [comment]
@wangyum What is the root cause?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 2]
```diff
@@ -188,6 +188,9 @@ private[hive] class IsolatedClientLoader(
...
+    name.startsWith("org.apache.derby.") ||
```
Comment (felixcheung) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: I'd also move this line to other org.apache* above

**Comment by jerryshao** [comment]
Please list out the reason why do you need such change? If it is a UT bug, why it didn't happen before?

**Comment by wangyum** [comment]
retest this please.

**Comment by holdenk** [comment]
So I had this come up while I was testing Spark 2.1.3 RC2 on a machine with an existing YARN cluster with `spark-testing-base`. Haven't had the chance to dig into it fully.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 3]
```diff
@@ -182,6 +182,7 @@ private[hive] class IsolatedClientLoader(
     name.startsWith("org.slf4j") ||
     name.startsWith("org.apache.log4j") || // log4j1.x
     name.startsWith("org.apache.logging.log4j") || // log4j2
+    name.startsWith("org.apache.derby.") ||
```
Comment (holdenk) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: So to clarify, this adds derby to the "shared" classes (we might want to mention this above on L139) and presumably this fix is for cases where we have an existing derby version on the class-path which may differ and if that happens we end up with strangeness on the metastore? cc @gatorsmile ?

Like I said, I've run into this issue so I'd like to see a fix :)

**Comment by gatorsmile** [comment]
LGTM

Thanks! Merged to master

**Comment by HyukjinKwon** [comment]
Sorry, why was this change required? I don't see https://github.com/apache/spark/pull/20944#issuecomment-379525776 is addressed Can you elaborate please? Why do we make `org.apache.derby` as shared?

Ideally, minor or maintenance versions of `derby` can be dumped up, and they shouldn't be shared unless there's a strong reason to keep it shared, for instance, making class resolution failed. How did you reproduce this and why the unit test is not added?

I found an actual issue while working on Apache Livy Spark 2.4 support. I am still investigating how it relates with the test failures but at the very list I see this specific commit matters since Apache Livy unittests pass without this specific commit.

Adding @vanzin

**Comment by HyukjinKwon** [comment]
Please describe manual tests and how it relates to actual usecase.

**Comment by HyukjinKwon** [comment]
ping @wangyum, I'm going to revert this today if there's no response today.

**Comment by wangyum** [comment]
Sorry @HyukjinKwon It's difficult reproduce. I am not sure whether it is caused by multithreading.
But you can verify it by:
https://github.com/apache/spark/blob/a75571b46f813005a6d4b076ec39081ffab11844/sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala#L117-L120

**Comment by HyukjinKwon** [comment]
I understood the reproducer step in JIRA but how and why it matters? Did it cause an actual problem in your production environment?

**Comment by HyukjinKwon** [comment]
To me, it sounds we made a fix because it was difficult to figure out exactly what's going on internally. It's okay if it's difficult to reproduce but it can be reproduce in production; however, the problem is that the fix of this caused another problem.

I am asking those to see why and how important this fix was.

**Comment by wangyum** [comment]
This fix for testing only, production won't use derby as their matestore database.

**Comment by HyukjinKwon** [comment]
Let's revert this then if this only targeted to fix the test. We can bring this back later when it's needed - tho, yea . This caused a specific case failure in Livy' when restarting Hive enabled Spark session.

@gatorsmile, I will revert this but I don't mind getting this back again if it actually fixes a usecase since I either don't know how exactly this fixes the case above and how it causes to make the Livy case failed - one thing clear is this specific commit is the cause.

Please revert me action if I missed an actual usecase fixed by this, I am okay.

**Comment by wangyum** [comment]
@HyukjinKwon I could reproduce this issue:
```shell
build/sbt clean package -Phive -Phive-thriftserver
export SPARK_PREPEND_CLASSES=true
bin/spark-sql --conf spark.sql.hive.metastore.version=2.3.4 --conf spark.sql.hive.metastore.jars=maven -e "create table t1 as select 1 as c"
```

**Comment by HyukjinKwon** [comment]
So this isn't test only fix anymore? If so, let's get it back. Let me check it too soon.

**Comment by HyukjinKwon** [comment]
Hm, but I still think https://github.com/apache/spark/pull/20944#issuecomment-436530497 is valid, actually. @wangyum, do we meet this issue when we use our binary releases, or did you meet this only in dev as described above?

**Comment by wangyum** [comment]
I hint this issue in two places:
1. Specify `spark.sql.hive.metastore.version` when using derby:
```
wget https://www-eu.apache.org/dist/spark/spark-2.4.1/spark-2.4.1-bin-hadoop2.7.tgz && tar -zxf spark-2.4.1-bin-hadoop2.7.tgz && cd spark-2.4.1-bin-hadoop2.7
bin/spark-sql --conf spark.sql.hive.metastore.version=2.0.0 --conf spark.sql.hive.metastore.jars=maven -e "create table t1 as select 1 as c"
```
2. Unit testing when the built-in Hive is upgraded to 2.3(https://github.com/apache/spark/pull/24044.  Remove [this line](https://github.com/apache/spark/pull/24044/files#diff-d613c921507243c65591c003a348f5f3R193) and run):
```  
build/sbt  "hive/testOnly *.HiveExternalSessionCatalogSuite *.HiveExternalCatalogSuite *.MultiDatabaseSuite"  -Phive -Phadoop-3.2
```

**Comment by HyukjinKwon** [comment]
I see. But can you check if Hive 3.1 support and onward still works with this change? My impression is that we will don't respect their derby version with this change anymore.

**Comment by wangyum** [comment]
The root cause is: Hive [creates more `java.sql.Connection`](https://github.com/apache/hive/blob/6cdfaaedd7d2b22fe32fa0f6686d3533012bf982/metastore/src/java/org/apache/hadoop/hive/metastore/txn/TxnHandler.java#L1760) since [`HIVE-10632(Hive 2.1.0)`](https://issues.apache.org/jira/browse/HIVE-10632).

**Pull request closed without merging**